src/transform/src/projection_pushdown.rs

```diff
@@ -0,0 +1,296 @@
+                        if let ScalarExpr::Column(c) = e {
+                            *c = *col_map.get(c).unwrap();
+                        }
+                    });
+                }
+                for c in projection.iter_mut() {
+                    if *c < input_arity {
+                        *c = *col_map.get(c).unwrap();
+                    } else {
+                        *c = *c - input_arity + input_proj.len();
+                    }
+                }
+                self.action(input, input_proj);
+                *relation = input
+                    .take_dangerous()
+                    .flat_map(func.clone(), exprs.clone())
+                    .project(projection);
+            }
+            RelationExpr::Map { input, scalars } => {
+                let input_arity = input.arity();
+                let mut needed: HashSet<_> = projection.iter().cloned().collect();
+                // First, prune any elements of scalars that are ignored by the projection.
+                // Compute the transitive closure of the needed columns.
+                let needed = loop {
+                    let mut new = needed.clone();
+                    for c in needed.iter() {
+                        if *c >= input_arity {
+                            new.extend(scalars[*c - input_arity].support());
+                        }
+                    }
+                    if new == needed {
+                        break new;
+                    }
+                    needed = new;
+                };
+                // First, prune away unneeded columns from the input.
+                let mut col_map = HashMap::new();
+                let mut input_proj = vec![];
+                for i in 0..input_arity {
+                    if needed.contains(&i) {
+                        col_map.insert(i, input_proj.len());
+                        input_proj.push(i);
+                    }
+                }
+
+                let mut new_scalars = vec![];
+                for (idx, expr) in scalars.iter_mut().enumerate() {
+                    if needed.contains(&(idx + input_arity)) {
+                        col_map.insert(idx + input_arity, input_proj.len() + new_scalars.len());
+                        expr.visit_mut(&mut |e| {
+                            if let ScalarExpr::Column(c) = e {
+                                *c = *col_map.get(c).unwrap();
+                            }
+                        });
+                        new_scalars.push(expr.clone());
+                    }
+                }
+                for c in projection.iter_mut() {
+                    *c = *col_map.get(c).unwrap();
+                }
+                self.action(input, input_proj);
+                *relation = input.take_dangerous().map(new_scalars).project(projection);
+            }
+            RelationExpr::Join {
+                inputs,
+                equivalences,
+                ..
+            } => {
```

Comment:
To look forward, there will come a moment when we turn this on that this fights with arrangement use. E.g. when we have an existing arrangement, but don't need all of its columns. We'll want to make sure that we are able to not confound the arrangement use, perhaps by re-absorbing any lingering projections into the join?

Reply:
We could try and use the same methodology for filters and maps too, maybe? Clearly we can do that if we know to, but another problem is that if we put some operators between the join and the arrangements, it becomes harder for everyone involved to realize that there is an opportunity. Just talking out loud at this point; not clear what to do!